src/main/java/org/opensearch/knn/index/codec/VectorSerializerFactory.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.io.ByteArrayInputStream;
+import java.io.ObjectStreamConstants;
+import java.util.Arrays;
+import java.util.Map;
+
+import static org.opensearch.knn.index.codec.VectorSerializerFactory.SerializationMode.ARRAY;
+import static org.opensearch.knn.index.codec.VectorSerializerFactory.SerializationMode.COLLECTION_OF_FLOATS;
+
+/**
+ * Class abstracts Factory for VectorSerializer implementations. Exact implementation constructed and returned based on
+ * either content of the byte array or directly based on serialization type.
+ */
+public class VectorSerializerFactory {
+    private static Map<SerializationMode, VectorSerializer> VECTOR_SERIALIZER_BY_TYPE = ImmutableMap.of(
+            ARRAY, new VectorAsArraySerializer(),
+            COLLECTION_OF_FLOATS, new VectorAsCollectionOfFloatsSerializer()
+    );
+
+    private static final int ARRAY_HEADER_OFFSET = 27;
+
+    private static final byte[] SERIALIZATION_PROTOCOL_HEADER_PREFIX = new byte[] {
+            highByte(ObjectStreamConstants.STREAM_MAGIC),
+            lowByte(ObjectStreamConstants.STREAM_MAGIC),
+            highByte(ObjectStreamConstants.STREAM_VERSION),
+            lowByte(ObjectStreamConstants.STREAM_VERSION),
+            ObjectStreamConstants.TC_ARRAY,
+            ObjectStreamConstants.TC_CLASSDESC
+    };
+
+    public static VectorSerializer getSerializerBySerializationMode(final SerializationMode serializationMode) {
+        return VECTOR_SERIALIZER_BY_TYPE.getOrDefault(serializationMode, new VectorAsCollectionOfFloatsSerializer());
```

Comment:
Im wondering if we throw exception if serializationMode isnt in VECTOR_SERIALIZER_BY_TYPE to prevent silent failure behavior.

Reply:
I think it's not necessary as we'll return  VectorAsCollectionOfFloatsSerializer implementation due to "OrDefault" part of the map.getOrDefault method. In other words new serialization method should be a default one.

Reply:
But is it ever expected that the serialization mode passed in does not map directly to a serializer or would this be a bug?

Reply:
This shouldn't be the case and to me it's a bug. My goal was to give us just enough potential for extensibility in future, as of today all serialization logic is hardcoded. It can be definitely reworked in future if needed.

Reply:
I hope this answers the question, resolving